sql: "invalid row length" in FK cascades code

Create the following logic test file:
```
statement ok
CREATE TABLE self_ab (a INT UNIQUE, b INT DEFAULT 1, INDEX (b))

statement ok
INSERT INTO self_ab VALUES (1, 2), (2, 2)

statement ok
ALTER TABLE self_ab ADD CONSTRAINT fk FOREIGN KEY (b) REFERENCES self_ab(a) ON UPDATE SET DEFAULT

statement ok
UPDATE self_ab SET a=3 WHERE a=2
```

About half the time I try to run this, I get the error below.

I spent some time trying to figure it out but there is a ton of code here that I'm not very familiar with. I would appreciate if someone more familiar with it can take a look.

```
panic: invalid row length 3, expected 2 [recovered]
	panic: panic while executing 1 statements: UPDATE _ SET _ = _ WHERE _ = _; caused by invalid row length 3, expected 2

goroutine 1142 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc00171ed80, 0x429dfe0, 0xc000cd5340, 0x349db00, 0xc00190a5a0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:731 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc00171ed80, 0x429dfe0, 0xc000cd5340)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0x61
panic(0x349db00, 0xc00190a5a0)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/rowcontainer.(*RowContainer).AddRow(0xc00040b560, 0x429dfe0, 0xc00152e440, 0xc00185ea50, 0x3, 0x3, 0x0, 0xc00106f700, 0x2, 0x2, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/rowcontainer/datum_row_container.go:221 +0x441
github.com/cockroachdb/cockroach/pkg/sql/row.(*cascader).cascadeAll(0xc0011b4600, 0x429dfe0, 0xc00152e440, 0xc001e2a480, 0xc00185f5f0, 0x2, 0x3, 0xc00185ea50, 0x3, 0x3, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/row/cascader.go:1019 +0x26ad
github.com/cockroachdb/cockroach/pkg/sql/row.(*Updater).UpdateRow(0xc000596a48, 0x429dfe0, 0xc00152e440, 0xc0016d6580, 0xc00185f5f0, 0x2, 0x3, 0xc00190a2c0, 0x1, 0x1, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/row/updater.go:490 +0x1c19
github.com/cockroachdb/cockroach/pkg/sql.(*tableUpdater).rowForUpdate(...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/tablewriter_update.go:51
github.com/cockroachdb/cockroach/pkg/sql.(*updateNode).processSourceRow(0xc000596a00, 0x429dfe0, 0xc00152e440, 0xc001603d40, 0xc00171f0f0, 0xc00185f5f0, 0x3, 0x3, 0x37f0600, 0xc000fb9801)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/update.go:328 +0x83d
github.com/cockroachdb/cockroach/pkg/sql.(*updateNode).BatchedNext(0xc000596a00, 0x429dfe0, 0xc00152e440, 0xc001603d40, 0xc00171f0f0, 0x3552200, 0xc000577470, 0xc000fb9880)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/update.go:186 +0x150
github.com/cockroachdb/cockroach/pkg/sql.(*rowCountNode).startExec(0xc0002ca440, 0x429dfe0, 0xc00152e440, 0xc001603d40, 0xc00171f0f0, 0xc00190a450, 0xc000fb98c8)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/plan_batch.go:159 +0xcd
github.com/cockroachdb/cockroach/pkg/sql.startExec.func2(0x39db8f8, 0x5, 0x429fba0, 0xc0002ca440, 0xc0011b48a0, 0xc000fb98d8)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:339 +0x55
```

--- a/pkg/sql/logictest/testdata/logic_test/cascade
+++ b/pkg/sql/logictest/testdata/logic_test/cascade
@@ -3707,13 +3707,11 @@ DROP TABLE c, b, a;
 # modifies the original table (#42117).
 subtest SelfReferencingCheckFail
 
-# TODO(radu): remove the FAMILY when #42120 is fixed.
 statement ok
 CREATE TABLE self_ab (
   a INT UNIQUE,
   b INT DEFAULT 1 CHECK (b != 1),
-  INDEX (b),
-  FAMILY (a, b)
+  INDEX (b)
 )
 
 statement ok
@@ -3739,3 +3737,43 @@ ALTER TABLE self_ab ADD CONSTRAINT fk2 FOREIGN KEY (a) REFERENCES self_ab_parent
 
 statement error failed to satisfy CHECK constraint \(b != 1\)
 UPDATE self_ab_parent SET p = 3 WHERE p = 2
+
+# Clean up.
+statement ok
+DROP TABLE self_ab, self_ab_parent
+
+# Extra test for self referencing foreign keys with set default.
+subtest SelfReferencingSetDefault
+
+statement ok
+CREATE TABLE self_abcd (
+  a INT DEFAULT 2,
+  b INT DEFAULT 2,
+  c INT DEFAULT 2,
+  d INT DEFAULT 2,
+  INDEX (c),
+  INDEX (d),
+  PRIMARY KEY (a), FAMILY (a, b, c, d)
+)
+
+statement ok
+INSERT INTO self_abcd VALUES (1, 2, 3, 4), (4, 1, 2, 3), (3, 4, 1, 2), (2, 3, 4, 1)
+
+statement ok
+ALTER TABLE self_abcd ADD CONSTRAINT fk1 FOREIGN KEY (c) REFERENCES self_abcd(a) ON UPDATE SET DEFAULT;
+ALTER TABLE self_abcd ADD CONSTRAINT fk2 FOREIGN KEY (d) REFERENCES self_abcd(a) ON UPDATE SET DEFAULT
+
+statement ok
+UPDATE self_abcd SET a = 5 WHERE a = 1
+
+query IIII
+SELECT * FROM self_abcd ORDER BY (a, b, c, d)
+----
+2  3  4  2
+3  4  2  2
+4  1  2  3
+5  2  3  4
+
+# Clean up.
+statement ok
+DROP TABLE self_abcd

--- a/pkg/sql/row/cascader.go
+++ b/pkg/sql/row/cascader.go
@@ -672,6 +672,7 @@ func (c *cascader) updateRows(
 	match sqlbase.ForeignKeyReference_Match,
 	values cascadeQueueElement,
 	action sqlbase.ForeignKeyReference_Action,
+	fk *sqlbase.ForeignKeyConstraint,
 	traceKV bool,
 ) (*rowcontainer.RowContainer, *rowcontainer.RowContainer, map[sqlbase.ColumnID]int, int, error) {
 	// Create the span to search for index values.
@@ -755,6 +756,28 @@ func (c *cascader) updateRows(
 	// values being updated will change based on both the original and updated
 	// values.
 	for i := values.startIndex; i < values.endIndex; i++ {
+		// If the rows are the same on the FK columns, then we don't want to continue the cascade.
+		// We skip this check when values.updatedValues is nil. This happens when a cascade is started
+		// because of a delete, upon which a row container is not created because there is not an
+		// updated value when a row is deleted.
+		// TODO (rohany): this check could be sped up greatly by storing a mask/map of what
+		//  actually changed instead of having to check every FK column.
+		if values.updatedValues != nil {
+			allFKColsUnchanged := true
+			origValues := values.originalValues.At(i)
+			updatedValues := values.updatedValues.At(i)
+			for _, colID := range fk.ReferencedColumnIDs {
+				colIdx := values.colIDtoRowIndex[colID]
+				if origValues[colIdx].Compare(c.evalCtx, updatedValues[colIdx]) != 0 {
+					allFKColsUnchanged = false
+					break
+				}
+			}
+			if allFKColsUnchanged {
+				continue
+			}
+		}
+
 		// Extract a single value to update at a time.
 		req, valueColIDtoRowIndex, err := batchRequestForIndexValues(
 			ctx, referencedIndex, referencingTable, referencingIndex, match, cascadeQueueElement{
@@ -1099,6 +1122,7 @@ func (c *cascader) cascadeAll(
 						sqlbase.ForeignKeyReference_SIMPLE,
 						elem,
 						foundFK.OnDelete,
+						foundFK,
 						traceKV,
 					)
 					if err != nil {
@@ -1130,6 +1154,7 @@ func (c *cascader) cascadeAll(
 						sqlbase.ForeignKeyReference_SIMPLE,
 						elem,
 						foundFK.OnUpdate,
+						foundFK,
 						traceKV,
 					)
 					if err != nil {

--- a/pkg/sql/row/updater.go
+++ b/pkg/sql/row/updater.go
@@ -184,7 +184,6 @@ func makeUpdaterWithoutCascader(
 		UpdateColIDtoRowIndex: updateColIDtoRowIndex,
 		primaryKeyColChange:   primaryKeyColChange,
 		marshaled:             make([]roachpb.Value, len(updateCols)),
-		newValues:             make([]tree.Datum, len(tableCols)),
 	}
 
 	if primaryKeyColChange {
@@ -265,6 +264,11 @@ func makeUpdaterWithoutCascader(
 		}
 	}
 
+	// If we are fetching from specific families, we might get
+	// less columns than in the table. So we cannot assign this to
+	// have length len(tableCols).
+	ru.newValues = make(tree.Datums, len(ru.FetchCols))
+
 	if checkFKs == CheckFKs {
 		var err error
 		if primaryKeyColChange {
@@ -287,13 +291,12 @@ func makeUpdaterWithoutCascader(
 // The return value is only good until the next call to UpdateRow.
 func (ru *Updater) UpdateRow(
 	ctx context.Context,
-	b *client.Batch,
+	batch *client.Batch,
 	oldValues []tree.Datum,
 	updateValues []tree.Datum,
 	checkFKs checkFKConstraints,
 	traceKV bool,
 ) ([]tree.Datum, error) {
-	batch := b
 	if ru.cascader != nil {
 		batch = ru.cascader.txn.NewBatch()
 	}
@@ -405,7 +408,7 @@ func (ru *Updater) UpdateRow(
 	}
 
 	// Add the new values.
-	ru.valueBuf, err = prepareInsertOrUpdateBatch(ctx, b,
+	ru.valueBuf, err = prepareInsertOrUpdateBatch(ctx, batch,
 		&ru.Helper, primaryIndexKey, ru.FetchCols,
 		ru.newValues, ru.FetchColIDtoRowIndex,
 		ru.marshaled, ru.UpdateColIDtoRowIndex,
@@ -480,7 +483,7 @@ func (ru *Updater) UpdateRow(
 	putFn := insertInvertedPutFn
 	// We're adding all of the inverted index entries from the row being updated.
 	for i := len(ru.Helper.Indexes); i < len(newSecondaryIndexEntries); i++ {
-		putFn(ctx, b, &newSecondaryIndexEntries[i].Key, &newSecondaryIndexEntries[i].Value, traceKV)
+		putFn(ctx, batch, &newSecondaryIndexEntries[i].Key, &newSecondaryIndexEntries[i].Value, traceKV)
 	}
 
 	if ru.cascader != nil {

